service/smithyprototype/lexruntimeservice/serializers.go

```diff
@@ -0,0 +1,542 @@
+
+	return next.HandleSerialize(ctx, in)
+}
+
+type awsRestJson1_serializeOpPostText struct{}
+
+func (p awsRestJson1_serializeOpPostText) ID() string {
+	return "awsRestJson1_serializeOpPostText"
+}
+
+func (p awsRestJson1_serializeOpPostText) HandleSerialize(ctx context.Context, in middleware.SerializeInput, next middleware.SerializeHandler) (
+	out middleware.SerializeOutput, metadata middleware.Metadata, err error,
+) {
+	request, ok := in.Request.(*smithyHTTP.Request)
+	if !ok {
+		return out, metadata, &aws.SerializationError{Err: fmt.Errorf("unknown transport type %T", in.Request)}
+	}
+
+	input, ok := in.Parameters.(*PostTextInput)
+	if !ok {
+		return out, metadata, &aws.SerializationError{Err: fmt.Errorf("unknown input parameters type %T", in.Parameters)}
+	}
+
+	restEncoder := rest.NewEncoder(request.Request)
+
+	restEncoder.AddHeader("Content-Type").String("application/json")
+
+	if err := awsRestJson1_serializeRestPostTextInput(input, restEncoder); err != nil {
+		return out, metadata, &aws.SerializationError{Err: err}
+	}
+
+	if err := restEncoder.Encode(); err != nil {
+		return out, metadata, &aws.SerializationError{Err: err}
+	}
+
+	jsonEncoder := json.NewEncoder()
+	if err := awsRestJson1_serializeJsonPostTextInput(input, jsonEncoder.Value); err != nil {
+		return out, metadata, &aws.SerializationError{Err: err}
+	}
+
+	if request, err = request.SetStream(bytes.NewReader(jsonEncoder.Bytes())); err != nil {
+		return out, metadata, &aws.SerializationError{Err: err}
+	}
+	in.Request = request
+
+	return next.HandleSerialize(ctx, in)
+}
+
+type awsRestJson1_serializeOpPutSession struct{}
+
+// ID is the middleware identifier
+func (p *awsRestJson1_serializeOpPutSession) ID() string {
+	return "awsRestJson1_serializeOpPutSession"
+}
+
+func (p *awsRestJson1_serializeOpPutSession) HandleSerialize(ctx context.Context, in middleware.SerializeInput, next middleware.SerializeHandler) (
+	out middleware.SerializeOutput, metadata middleware.Metadata, err error,
+) {
+	request, ok := in.Request.(*smithyHTTP.Request)
+	if !ok {
+		return out, metadata, &aws.SerializationError{Err: fmt.Errorf("unknown transport type %T", in.Request)}
+	}
+
+	input, ok := in.Parameters.(*PutSessionInput)
+	if !ok {
+		return out, metadata, &aws.SerializationError{Err: fmt.Errorf("unknown input parameters type %T", in.Parameters)}
+	}
+
+	restEncoder := rest.NewEncoder(request.Request)
+
+	restEncoder.AddHeader("Content-Type").String("application/json")
+
+	if err := awsRestJson1_serializeRestPutSessionInput(input, restEncoder); err != nil {
+		return out, metadata, &aws.SerializationError{Err: err}
+	}
+
+	jsonEncoder := json.NewEncoder()
+	if err := awsRestJson1_serializeJsonPutSessionInput(input, jsonEncoder.Value); err != nil {
+		return out, metadata, &aws.SerializationError{Err: err}
+	}
+
+	if request, err = request.SetStream(bytes.NewReader(jsonEncoder.Bytes())); err != nil {
+		return out, metadata, &aws.SerializationError{Err: err}
+	}
```

Comment:
Should `json.NewEncoder` take a buffer to write to instead of having its own internal buffer?  This way that buffer can be sent to set stream directly without creating the new `bytes.Reader`?  This would also allow us to write to any underlying io.Writer, making the Encoder more flexible, and code gen could pre-allocate estimated size preventing multiple future allocations during serialization.

```go
// could optimistically pre-allocate a buffer if we have rough idea of size, prevent multiple
// allocations later.
streamBuf := bytes.NewBuffer(nil)

jsonEncoder := json.NewEncoder(streamBuf)
if err := awsRestJson1_serializeJsonPutSessionInput(input, jsonEncoder.Value); err != nil {
	return out, metadata, &aws.SerializationError{Err: err}
}

if request, err = request.SetStream(streamBuf); err != nil {
	return out, metadata, &aws.SerializationError{Err: err}
}
```

Reply:
This feedback concerns serializer implementation. @skmcgrail can you take a look?